docs/device-emulation.md

```diff
@@ -0,0 +1,59 @@
+---
+title: Device Emulation
+---
+
+## Introduction
+
+Using device-tests, fwupd can prevent regressions by updating and downgrading firmware on real
+hardware. However, much past a few dozen devices this does not scale, either by time, or because
+real devices need plugging in and out. We can unit test the plugin internals, but this does not
+actually test devices being attached, removed, and being updated.
+
+By recording the backend devices we can build a "history" in GUsb of what control, interrupt and
+bulk transfers were sent to, and received from the device. By dumping these we can "replay" the
+update without the physical hardware connected.
+
+There are some problems that make this emulation slightly harder than the naive implementation:
+
+* Devices are sometimes detached into a different "bootloader" device with a new VID:PID
+* Devices might be "composite" and actually be multiple logical devices in one physical device
+* Devices might not be 100% deterimistic, e.g. queries might be processed out-of-order
+
+For people to generate and consume emulated devices, we do need to make the process easy to
+understand, and also easy to use. Some key points that we think are important, is the ability to:
+
+* Dump the device of an unmodified running daemon.
+* Filter to multiple or single devices, to avoid storing data for unrelated parts of the system.
+* Load an emulated device into an unmodified running daemon.
+
+Because we do not want to modify the daemon, we think it makes sense to *load* and *save* emulation
+state over D-Bus. Each phase can be controlled, which makes it easy to view, and edit, the recorded
+emulation data.
+
+For instance, calling `fwupdmgr emulate-add` would ask the end user to choose a device to start
```

Comment:
`emulate-add` may be confusing, it sound similar to `emulate-load` to me, what about `emulate-select`?

Reply:
I think it's also the `allow-emulate-save` that's a problem. I think the flag name should relate to the fwupdmgr command too.

Reply:
Okay, we went with `emulation-tag` --fixed.